pkg/bpf/migrate/migrate.go

```diff
@@ -0,0 +1,376 @@
+}
+
+// Start starts the maps migration process for a specific elf file's
+// maps. It returns true if it created a pending file and false if it
+// didn't do anything (i.e. the pinned/old map is the same as the new
+// one).
+// The "syslogger" argument is optional
+func Start(pathName string, syslogger *syslog.Writer) (bool, error) {
+	coll, err := ebpf.LoadCollectionSpec(pathName)
+	if err != nil {
+		return false, err
+	}
+	var moved bool
+	for n, m := range coll.Maps {
+		var pinning uint32
+		// discard the id, it is not needed
+		if _, err := io.CopyN(io.Discard, &m.Extra, 4); err != nil {
+			return false, fmt.Errorf("could not read id field of map %q: %v", n, err)
+		}
+		if err := binary.Read(&m.Extra, coll.ByteOrder, &pinning); err != nil {
+			return false, fmt.Errorf("could not read pinning field of map %q: %v", n, err)
+		}
+		if pinning != pinGlobalNS {
+			continue
+		}
+		mv, err := bpfPending(n, m, syslogger)
+		if err != nil {
+			return moved, err
+		}
+		if mv {
+			moved = true
+		}
+	}
+	return moved, nil
+}
+
+func bpfPending(name string, m *ebpf.MapSpec, syslogger *syslog.Writer) (bool, error) {
+	file := fmt.Sprintf("%s/%s", fsBase, name)
+	_, err := os.Stat(file)
+	if err != nil {
+		if os.IsNotExist(err) {
+			return false, nil
+		}
+		return false, fmt.Errorf("cannot stat node %q", file)
+	}
+	pinned, err := ebpf.LoadPinnedMap(file, nil)
+	if err != nil {
+		return false, err
+	}
+	if pinned.Type() == m.Type &&
+		pinned.KeySize() == m.KeySize &&
+		pinned.ValueSize() == m.ValueSize &&
+		pinned.Flags() == m.Flags &&
+		pinned.MaxEntries() == m.MaxEntries {
+		return false, nil
+	}
+	dest := fmt.Sprintf("%s:%s", file, statePending)
+	if syslogger != nil {
+		syslogger.Warning(fmt.Sprintf("Property mismatch in %s, migrating node to %s!\n", file, dest))
+	}
+	err = syscall.UtimesNano(file, nilTimeSpecArg)
+	if err != nil {
+		return false, fmt.Errorf("could not update timestamp of %q", file)
+	}
+	err = os.Rename(file, dest)
+	if err != nil {
+		return false, err
+	}
+	return true, nil
+}
+
+// Finish finishes migrating the maps of a specific elf file's maps.
+func Finish(pathName string, exit int, syslogger *syslog.Writer) error {
+	coll, err := ebpf.LoadCollectionSpec(pathName)
+	if err != nil {
+		return err
+	}
+	for n := range coll.Maps {
+		err := bpfFinalize(n, exit, syslogger)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func bpfFinalize(name string, exit int, syslogger *syslog.Writer) error {
+	file := fmt.Sprintf("%s/%s:%s", fsBase, name, statePending)
+	_, err := os.Stat(file)
+	if err != nil {
+		if os.IsNotExist(err) {
+			return nil
+		}
+		return fmt.Errorf("cannot stat node %q", file)
+	}
+	if exit != 0 {
+		dest := fmt.Sprintf("%s/%s", fsBase, name)
+		if syslogger != nil {
+			syslogger.Warning(fmt.Sprintf("Restoring migrated node %s into %s due to bad exit.\n", file, dest))
+		}
+		err = syscall.UtimesNano(file, nilTimeSpecArg)
```

Comment:
Same as my earlier comment, why do acc and mod timestamps need to be bumped? Is this something we should incorporate into `Map.Pin()`?

Reply:
Potentially. I'd be interested in a more comprehensive from @borkmann .